data/users.js

```diff
@@ -0,0 +1,170 @@
+  assertIsValuedString,
+  assertRequiredObject,
+  assertRequiredNumber,
+} = require("../utils/assertion");
+
+const getByObjectId = async (objectId) => {
+    const collection = await getUserCollection();
+    const user = await collection.findOne(idQuery(objectId));
+    return parseMongoData(user);
+};
+
+const getAllUsers = async () => {
+    const collection = await getUserCollection();
+
+    const userList = await collection.find({}).toArray();
+
+    return parseMongoData(userList);
+},
+  
+const getUser = async (id) => {
+    assertObjectIdString(id);
+    return await getByObjectId(new ObjectId(id));
+};
+
+const createUser = async (data) => {
+    assertRequiredObject(data);
+  
+    const { corporateId, rankId, name, password, email, contact, designation, rank, role, createdBy, createdAt = new Date().getTime() } = data;
+  
+    assertObjectIdString(corporateId, "Corporate ID");
+    assertObjectIdString(rankId, "Rank ID");    
+    assertIsValuedString(name, "User name");
+    assertIsValuedString(password, "Password");
+    assertIsValuedString(email, "Email");
+    assertRequiredNumber(contact, "Contact Number");
+    assertIsValuedString(designation, "Designation");
+    assertIsValuedString(role, "Role");
+    assertRequiredNumber(rank, "Rank");
+    assertIsValuedString(createdBy, "Created By");
+    assertRequiredNumber(createdAt, "User created time");
+  
+    const userData = {
+      _id: new ObjectId(),
+      corporateId: new ObjectId(corporateId),
+      rankId: new ObjectId(rankId),
+      name : name,
+      password: password,
+      email: email,
+      contact: contact,
+      designation: designation,
+      rank: rank,
+      role: role,
+      createdBy: createdBy,
+      createdAt: createdAt,
+      updatedAt: createdAt,
+    };
+  
+    const collection = await getUserCollection();
+    const { result, insertedCount, insertedId } = await collection.insertOne(
+      userData
+    );
+  
+    if (!result.ok || insertedCount !== 1) {
+      throw new QueryError(`Could not create user for corporate ID(${corporateId})`);
+    }
+  
+    return await getByObjectId(insertedId);
+};
+
+const updateRank = async (id, updates) => {
+    assertObjectIdString(id);
+    assertRequiredObject(updates, "Ranks updates data");
+
+    const { corporateId, rankId, name, password, email, contact, designation, rank, role, createdBy, createdAt = new Date().getTime() } = updates;
+  
+    assertObjectIdString(corporateId, "Corporate ID");
+    assertObjectIdString(rankId, "Rank ID");    
+    assertIsValuedString(name, "User name");
+    assertIsValuedString(password, "Password");
+    assertIsValuedString(email, "Email");
+    assertRequiredNumber(contact, "Contact Number");
+    assertIsValuedString(designation, "Designation");
+    assertIsValuedString(role, "Role");
+    assertRequiredNumber(rank, "Rank");
+    assertIsValuedString(createdBy, "Created By");
+    assertRequiredNumber(createdAt, "User created time");
+  
```

Comment:
instead of createdBy/At, I suppose it's updatedBy/At you want?

You might need to pass in the `userId` for the one who is updating(who's calling it from the routes)
`assertObjectIdString(userId)`
`assertRequiredNumber(udpatedAt)`